torch/csrc/autograd/input_metadata.h

```diff
@@ -67,9 +107,59 @@ struct InputMetadata {
     return at::zeros(shape_, options_);
   }
 
+  at::IntArrayRef compute_shape(const at::Tensor& t){
+    if(t.is_nested()){
+      return at::IntArrayRef{};
+    }
+    return t.sizes();
+  }
+
+  c10::optional<const at::Tensor> compute_nested_shape_from_tensor(const at::Tensor& t){
+    if(t.is_nested()){
+      auto nested_impl = at::native::get_nested_tensor_impl(t);
+      return nested_impl->get_nested_size_tensor();
+    }
+    return c10::nullopt;
+  }
+
+  bool is_same_shape(const at::Tensor& grad) const {
+    TORCH_CHECK(
+        is_nested_tensor() == grad.is_nested(),
+        "Metadata and grad both need to be not nested or nested tensors.")
+    if (is_nested_tensor()) {
+      return at::native::get_nested_size_tensor(grad).equal(nested_shape());
+    } else {
+      return grad.sizes().equals(shape());
+    }
+  }
+  bool is_expandable_to_shape(const at::Tensor& grad) const{
+    // TODO: Currently NestedTensors are not expandable.
+    return is_nested_tensor()? false: at::is_expandable_to(shape(), grad.sizes());
+  }
+
+  std::stringstream incompatible_shape_error_message(
+      const size_t index,
+      const at::Tensor& grad) const {
+    std::stringstream ss;
+    ss << "invalid gradient at index " << index << " - got ";
+    if (grad.is_nested()) {
+      ss << at::native::get_nested_size_tensor(grad);
+    } else {
+      ss << grad.sizes();
+    }
+    ss << " but expected shape compatible with ";
+    if (is_nested_tensor()) {
+      ss << nested_shape();
+    } else {
+      ss << shape();
+    }
+    return ss;
+  }
+
 private:
   const at::TensorOptions options_;
   at::DimVector shape_;
```

Comment:
How about this: Let's turn this into a union type? If it's a regular Tensor it'll stay a DimVector, otherwise we'll turn it into a nested size.